Change/add to current functions;  BlueBackPixel.java

gwarby opened the pull request:
Changed servo functions so wait is optional, if used, only waits until servo reaches position +5 ms. Implemented change, added middle/right programs. All need tuning on field

gwarby committed:
Change/add to current functions;  BlueBackPixel.java

Changed servo functions so wait is optional, if used, only waits until servo reaches position +5ms. Implemented change, added middle/right programs. All need tuning on field

gwarby committed:
Merge branch 'main' into Finishing-autonomous

gwarby committed:
Bug fix BlueBackPixel.java

Changed wait in ms to "wait" string on line 266

gwarby committed:
Create RedBackPixel.java with extra time save

Created as a copy of BlueBackPixel.java, reversed all necessary numbers, estimated movements, and changed program so the robot backs up the whole way to the backdrop as opposed to doing a 180 upon arrival. Hopefully a time saver. Used time save functions recently implemented in BlueBackPixel

lcdnbl commented:
@gwarby The following documentation specs that `getPosition()` simply returns the last _commanded_ position.  I think your idea is superior to the manual wait times, but I'm not sure servos have a way to wait for "not busy" or "position ~= x".
[ServoImpl:getPosition()](https://ftctechnh.github.io/ftc_app/doc/javadoc/com/qualcomm/robotcore/hardware/ServoImpl.html#getPosition--)

gwarby committed:
Redo custom wait time BlueBackPixel.java

Calculate distance servo has to move, move servo, sleep length dependent on distance. There are separate, constant factors for grabber and flipper servos

gwarby committed:
Parity update RedBackPixel.java 

Changed to new custom wait times, implemented through program

gwarby committed:
2nd BlueBackPixel.java custom wait change

Implemented new function name through code

lcdnbl commented on BlueBackPixel.java:
```diff
@@ -537,40 +539,28 @@ private void armextend(double extend_retract, double power) {
+  private void closeClamp() { grabber.setPosition(GRABBER_SERVO_CLOSED_POS); }
+  private void closeClampWait() {
+    double distance = Math.abs(grabber.getPosition - GRABBER_SERVO_CLOSED_POS); // Calculate movement BEFORE setting new position
+    grabber.setPosition(GRABBER_SERVO_OPENED_POS); // Set new position
```
We found and fixed this issue (_OPENED_ on line 547 should be _CLOSED_ like the rest of this function) at Tuesday's team meeting.

lcdnbl commented on BlueBackPixel.java:
```diff
@@ -579,33 +569,25 @@ private void openClampLittle(string wait) {
-    }
-    sleep(5);
+  private void reverseFlipperWait() {
+    double distance = Math.abs(grabber.getPosition - CLAW_FLIP_SERVO_FLIPPED_POS);
+    grabber.setPosition(CLAW_FLIP_SERVO_FLIPPED_POS);
+    sleep(distance * CALC_FLIPPER_WAIT_MS + 5);
```
@gwarby I'm not sure if you caught & fixed this during Tuesday's meeting, but this Flipper function has grabber calls in it.

lcdnbl commented on BlueBackPixel.java:
```diff
@@ -579,33 +569,25 @@ private void openClampLittle(string wait) {
-        sleep(1);
-      }
-      sleep(5);
+  private void normalFlipperWait() {
+    double distance = Math.abs(grabber.getPosition - CLAW_FLIP_SERVO_NORMAL_POS);
+    grabber.setPosition(CLAW_FLIP_SERVO_NORMAL_POS);
+    sleep(distance * CALC_FLIPPER_WAIT_MS + 5);
```
@gwarby this one also has grabber instead of flipper in the get/set calls

lcdnbl commented on BlueBackPixel.java:
```diff
@@ -579,33 +569,25 @@ private void openClampLittle(string wait) {
-      sleep(5);
-    }
+  private void groundTransitionFlipperWait() {
+    double distance = Math.abs(grabber.getPosition - CLAW_FLIP_SERVO_TO_FROM_GROUND);
+    grabber.setPosition(CLAW_FLIP_SERVO_TO_FROM_GROUND);
```
@gwarby here too
